Disable compatibility layer via hook

gigitux opened the pull request:
This PR introduces a new filter called `woocommerce_disable_compatibility_layer` that allows disabling the compatibility layer.
At the same time, using the filter, the compatibility layer is disabled when Archive Product templates and Single Product template have the WooCommerce Product Grid Block.

<!-- Reference any related issues or PRs here -->

Fixes #8505 

### Testing
#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

##### Archive Templates

1. With the [Twenty Twenty-Three](https://wordpress.org/themes/twentytwentythree/) theme installed, go to `Appearance » Editor`.
2. Edit the Product Catalog template, adding the Products block above the WooCommerce Product Grid Block.
3. Save.
4. Go to the front end and verify that the `WooCommerce Product Grid Block` is visible.



* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Add the `woocommerce_disable_compatibility_layer` filter to disable the compatibility layer. Disable the compatibility layer when Archive Product and Single Product templates contain the WooCommerce Product Grid Block.

gigitux committed:
Disable compatibility layer via hook

Aljullu commented on src/Templates/BlockTemplatesCompatibility.php:
```diff
@@ -39,8 +39,52 @@ protected function init() {
+			'render_block_data',
+			function( $parsed_block, $source_block, $parent_block ) {
+				/**
+				* Filters to disable the compatibility layer for the blockified templates.
+				*
+				* This hooks allows to disable the compatibility layer for the blockified.
```
```suggestion
				* Filter to disable the compatibility layer for the blockified templates.
				*
				* This hook allows to disable the compatibility layer for the blockified templates.
```

Aljullu commented on src/BlockTemplatesController.php:
```diff
@@ -5,6 +5,7 @@
 use Automattic\WooCommerce\Blocks\Templates\BlockTemplatesCompatibility;
 use Automattic\WooCommerce\Blocks\Templates\ProductAttributeTemplate;
 use Automattic\WooCommerce\Blocks\Utils\BlockTemplateUtils;
+use Automattic\WooCommerce\Blocks\Templates\ProductSearchResultsTemplate;
```
Is this class used somewhere in this file? If not, do we need to `use` it?

gigitux replied:
Removed!

Aljullu reviewed (approved):
One question and a suggestion to update some texts, but besides that, LGTM and is testing like a charm. :ok_hand:

gigitux committed:
Update src/Templates/BlockTemplatesCompatibility.php

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

gigitux committed:
remove useless import

gigitux commented:
> One question and a suggestion to update some texts, but besides that, LGTM and is testing like a charm. 👌

Thanks for the suggestions! I addressed your feedback! I will merge it 🚀

gigitux committed:
Merge branch 'trunk' into fix/8505-disable-compatibility-layer
